CB-14656 Fix unbound service restart of existing stacks

Bajzathd opened the pull request:
The faulty unbound service restart (systemctl restart unbound) that was burned to 7.2.11 images caused issues on multiple stacks. Hotfix images were created containing the correct unbound service restart (pkill -u unbound -SIGHUP unbound), and the purpose of this commit is to fix the stacks created with the faulty images.
Implementation details:
- Create a new salt state that fixes the command, that will be applied to new stacks: faulty_7_2_11_images_unbound_restart_patch
- Introduce a new quartz job that runs for existing stacks: ExistingStackPatcherJob
- Trigger a full salt update on the affected stacks to upload and apply the new salt file to existing stacks: UnboundRestartPatchService

pdarvasi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/UnboundRestartPatchService.java:
```diff
@@ -0,0 +1,107 @@
+            return affected;
+        } catch (CloudbreakImageNotFoundException e) {
+            LOGGER.warn("Image not found for stack " + stack.getResourceCrn(), e);
+            throw new IllegalStateException("Image not found for stack " + stack.getResourceCrn(), e);
```
is there any specific reason to throw this `IllegalStateException` here?

bbihari replied:
CloudbreakRuntimeException?

Bajzathd replied:
Changed to CloudbreakRuntimeException

lnardai commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherJob.java:
```diff
@@ -0,0 +1,112 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExistingStackPatcherJob.class);
+
+    private static final EnumSet<Status> FAILED_OR_IN_DELETE_STATUSES = EnumSet.of(
```
STOP state should also be filtered right?

Bajzathd replied:
No, these are the statuses where we will not even retry the patching. 
Stop case will be handled by whether the CM server is in reachable state, so if the stack is stopped, we will try to patch it in the next job run.

pdarvasi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/UnboundRestartPatchService.java:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+
+    private boolean isCmServerReachable(Stack stack) throws ExistingStackPatchApplyException {
```
this logic seems to belong to another CM/instancemetadata related service for future reuse

Bajzathd replied:
Moved to `stack.getClusterManagerServer()`

pdarvasi replied:
Thanks, the whole `isCmServerReachable` logic should be reusable, but it could be refactored when needed.

lnardai commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherJob.java:
```diff
@@ -0,0 +1,112 @@
+    protected void executeTracedJob(JobExecutionContext context) throws JobExecutionException {
+        Stack stack = stackService.getByIdWithListsInTransaction(getStackId());
+        Status stackStatus = stack.getStatus();
+        if (!FAILED_OR_IN_DELETE_STATUSES.contains(stackStatus)) {
```
This would rather be active or running clusters right?

lnardai replied:
What if there is an ongoing Scale or Upgrade operation? Shouldn't there be a check for no Flow running for an operation? How long does the stack patch takes?

lnardai commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherJobInitializer.java:
```diff
@@ -0,0 +1,32 @@
+    public void initJobs() {
+        if (config.isExistingStackPatcherEnabled()) {
+            LOGGER.info("Scheduling stack patcher jobs");
+            getAliveAndNotDeleteInProgressStacksStream()
```
I see a logic where we are skipping the StackPatch operation if it was already applied, but we still schedule jobs for them until the feature is disabled?

pdarvasi commented on core/src/main/resources/schema/app/20211103115000_CB-14656_create_stackpatch_table.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE TABLE IF NOT EXISTS stackpatch
+(
+    id              bigint NOT NULL,
+    type            CHARACTER VARYING(255) NOT NULL,
```
maybe an index to match `findByStackAndType` would be useful

Bajzathd replied:
I agree, adding it

pdarvasi commented on orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java:
```diff
@@ -819,6 +819,17 @@ public void removeDeadSaltMinions(GatewayConfig gateway) throws CloudbreakOrches
     }
 
+    @Override
+    public Map<String, String> replacePatternInFileOnAllHosts(GatewayConfig gatewayConfig, String file, String pattern, String replace)
```
where is this called from?

Bajzathd replied:
It remained there from the previous implementation (file update instead of full salt update), I think we can leave it there

Bajzathd commented on core-model/src/main/java/com/sequenceiq/cloudbreak/domain/stack/StackPatchType.java:
```diff
@@ -0,0 +1,6 @@
+
+public enum StackPatchType {
+    UNBOUND_RESTART,
+    UNKNOWN
```
This is the fallback value for the enum converter - so that if we already save an enum value to the db but we have to roll back to a CB version that does not support that new enum value, it will not fail. See `StackPatchTypeConverter`.

Bajzathd commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherConfig.java:
```diff
@@ -0,0 +1,33 @@
+package com.sequenceiq.cloudbreak.job.existingstackfix;
```
Nice catch, fixing it

bbihari commented on core/src/main/java/com/sequenceiq/cloudbreak/job/AbstractStackJobInitializer.java:
```diff
@@ -0,0 +1,34 @@
+
+    protected Stream<Stack> getAliveAndNotDeleteInProgressStacksStream() {
+        return getAliveStacksStream()
+                .filter(s -> !s.isStackInDeletionOrFailedPhase());
```
Btw. getAllAlive is a _very_ optimized call

bbihari commented on core/src/main/java/com/sequenceiq/cloudbreak/job/stackpatcher/ExistingStackPatcherConfig.java:
```diff
@@ -0,0 +1,33 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExistingStackPatcherConfig.class);
+
+    @Value("${existingstackpatcher.intervalhours}")
```
This should have a default value.

Bajzathd replied:
As discussed offline, the service will not start without providing a value so it can stay like that

bbihari commented on core/src/main/java/com/sequenceiq/cloudbreak/job/stackpatcher/ExistingStackPatcherConfig.java:
```diff
@@ -0,0 +1,33 @@
+    @Value("${existingstackpatcher.intervalhours}")
+    private int intervalInHours;
+
+    @Value("${existingstackpatcher.enabled}")
```
This too needs a default value

bbihari commented on core/src/main/java/com/sequenceiq/cloudbreak/job/stackpatcher/ExistingStackPatcherJobService.java:
```diff
@@ -0,0 +1,81 @@
+                .withIdentity(jobDetail.getKey().getName(), TRIGGER_GROUP)
+                .withDescription("Trigger for patching existing stack")
+                .withSchedule(SimpleScheduleBuilder.simpleSchedule()
+//                        .withIntervalInHours(properties.getIntervalInHours())
```
??? :)

lacikaaa commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/ExistingStackPatchService.java:
```diff
@@ -0,0 +1,63 @@
+        if (!flowLogService.isOtherFlowRunning(stack.getId())) {
+            try {
+                LOGGER.info("Starting stack {} patching for {}", stack.getResourceCrn(), getStackFixType());
+                checkedMeasure(() -> doApply(stack), LOGGER, "Existing stack patching {} took {} ms for stack: {}.", getStackFixType(), stack.getResourceCrn());
```
checkedmeasure will provide the first param if I remember well

pdarvasi replied:
`    private static void logDuration(Logger logger, String message, long start, Object[] params) {
        Object[] extParams = Stream.concat(Stream.of(System.currentTimeMillis() - start), Stream.of(params)).toArray();`

Bajzathd replied:
So, do I have any TODOs here? :smile:

lacikaaa replied:
I see :) so the first parameter will always be the duration, so you have to rephrase the log message like `took {} ms to patch {} for stack {}`

Bajzathd replied:
Nice catch, fixed it

pdarvasi replied:
Exactly ;)

lacikaaa commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/ExistingStackPatchService.java:
```diff
@@ -0,0 +1,63 @@
+                throw new ExistingStackPatchApplyException(message, e);
+            }
+        } else {
+            throw new ExistingStackPatchApplyException("Another flow is running for stack " + stack.getResourceCrn());
```
is the exception really necessary? I mean this is more like business as usual, and we should just log that it didn't happened because a flow was running

Bajzathd replied:
If we do not throw an exception, the job will be unscheduled. Another option would be to use a boolean to indicate if the apply was successful, would it be a better solution?

lacikaaa replied:
so in my opinion a better way would be if this `apply` method would return with a result object, which could have error(s) and success, and depending on that the caller side in `ExistingStackPatcherJob#applyStackPatches` could decide if it was successful. At least that how we collect successful and failed operations on FMS side. Exception is an expensive operation in Java world, and I think we already build too many logic on that, like salt polling. So if it's possible I would avoid throwing exceptions if it's not really an error. And a running flow in case of a cluster is just fine. (sure, you would throw an exception if this would be a request on the API, to notify the caller that the call failed.)
but if this would be a too big refactor we can leave this way for now and do it later

Bajzathd replied:
I see, thank you! I will include this optimization in the follow up ticket we discussed with @pdarvasi

Bajzathd commented on core/src/main/resources/schema/app/20211103115000_CB-14656_create_stackpatch_table.sql:
```diff
@@ -0,0 +1,26 @@
+
+ALTER TABLE ONLY stackpatch ADD CONSTRAINT uk_stackpatch_stack_type UNIQUE (stack_id, type);
+
+CREATE INDEX IF NOT EXISTS stackpatch_type_stack ON stackpatch(type, stack_id);
```
Nice catch, fixing it

pdarvasi commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherConfig.java:
```diff
@@ -0,0 +1,33 @@
+    private int intervalInHours;
+
+    @Value("${existingstackpatcher.enabled}")
+    private boolean existingStackPatcherEnabled;
```
Yes, pls file a follow-up!

Bajzathd replied:
Follow up created: https://jira.cloudera.com/browse/CB-14844

pdarvasi commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherJobService.java:
```diff
@@ -0,0 +1,80 @@
+                .withSchedule(SimpleScheduleBuilder.simpleSchedule()
+                        .withIntervalInHours(properties.getIntervalInHours())
+                        .repeatForever()
+                        .withMisfireHandlingInstructionNextWithRemainingCount())
```
yes, but it makes a little more sense to look at, so pls change it.

Bajzathd commented on core/src/main/java/com/sequenceiq/cloudbreak/job/existingstackfix/ExistingStackPatcherJob.java:
```diff
@@ -0,0 +1,112 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExistingStackPatcherJob.class);
+
+    private static final EnumSet<Status> FAILED_OR_IN_DELETE_STATUSES = EnumSet.of(
```
As discussed offline, the check will be extended to not start the stack patch if the last flow failed, this way the user can retry it.

akanto reviewed (changes requested):
Beyond the low-level test that has been added to this feature, please make sure that this feature is tested before it is merged. Kindly note that e2e tests do not provide any coverage for this, since clusters launched by the e2e tests are not using faulty images.

Bajzathd commented:
Last flow failed check was incorrect, so I had to modify it. I have refactored the FlowRetryService to have a check if we have any restartable flows.
        modified:   core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/ExistingStackPatchService.java
        modified:   core/src/test/java/com/sequenceiq/cloudbreak/service/stackpatch/ExistingStackPatchServiceTest.java
        new file:   flow/src/main/java/com/sequenceiq/flow/domain/RetryableState.java
        new file:   flow/src/main/java/com/sequenceiq/flow/domain/RetryableStateResponse.java
        modified:   flow/src/main/java/com/sequenceiq/flow/service/FlowRetryService.java
